### PR TITLE
webhook: pod mutating handler add extended resource spec

### PR DESCRIPTION
**Pull request by saintube**
Signed-off-by: saintube <saintube@foxmail.com>

### Ⅰ. Describe what this PR does

Update pod mutating webhook to add a pod annotation of extended resource spec.

### Ⅱ. Does this pull request fix one issue?

Partially for #717.

### V. Checklist

- [X] I have written necessary docs and comments
- [X] I have added necessary unit tests and integration tests
- [X] All checks passed in `make test`

### CODE REVIEW

File: pkg/webhook/pod/mutating/extended_resource_spec.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+func (h *PodMutatingHandler) mutateByExtendedResources(pod *corev1.Pod) error {
+	// dump batch-resource of pod.spec.containers[*].resources.requests/limits into ExtendedResourceSpec{}
+	extendedResourceSpec := &extension.ExtendedResourceSpec{
+		Containers: map[string]extension.ExtendedResourceContainerSpec{},
+	}
+
+	// TODO: count pod overhead
+	for i := range pod.Spec.InitContainers {
+		container := pod.Spec.InitContainers[i]
```
Comment (eahydra) [inline pkg/webhook/pod/mutating/extended_resource_spec.go:48]: here we can get container pointer.

File: pkg/webhook/pod/mutating/extended_resource_spec.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+			extension.BatchCPU,
+			extension.BatchMemory,
+		})
+		if r == nil {
+			continue
+		}
+		extendedResourceSpec.Containers[container.Name] = *r
+	}
+	for i := range pod.Spec.Containers {
+		container := pod.Spec.Containers[i]
```
Comment (eahydra) [inline pkg/webhook/pod/mutating/extended_resource_spec.go:59]: and here

File: pkg/webhook/pod/mutating/extended_resource_spec.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		return nil
+	}
+
+	// compare annotation values
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+
+	spec, err := extension.GetExtendedResourceSpec(pod.Annotations)
+	if err != nil {
```
Comment (eahydra) [inline pkg/webhook/pod/mutating/extended_resource_spec.go:81]: the function `GetExtendedResourceSpec`  will return nil, nil even if there is no extendResourceSpec annotation. Maybe we should return an empty extendResourceSpec object.

File: pkg/webhook/pod/mutating/extended_resource_spec.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	// no requirement of specified extended resources
+	if len(extendedResourceSpec.Containers) <= 0 {
+		return nil
+	}
+
+	// compare annotation values
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
```
Comment (eahydra) [inline pkg/webhook/pod/mutating/extended_resource_spec.go:77]: we can move the code  to line 90

File: pkg/webhook/pod/mutating/extended_resource_spec.go (hunk 1 above)
Comment (saintube) [reply pkg/webhook/pod/mutating/extended_resource_spec.go]: fixed

File: pkg/webhook/pod/mutating/extended_resource_spec.go (hunk 3 above)
Comment (saintube) [reply pkg/webhook/pod/mutating/extended_resource_spec.go:76]: done

**Review by eahydra — APPROVED** [review]
/lgtm

**Pull request merged**